### PR TITLE
Resize keep ratio

### PR DESCRIPTION
**Pull request by Sara-Eftekhar-Azam**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,7 @@
 code/data_analysis/dataset/
 code/data_analysis/rooftile_dataset_complete_rescaled/
-code/data_analysis/dataset_rooftile_small/
-code/data_analysis/dataset_rooftile_640/
+code/data_analysis/dataset_rooftile_small/**
+code/data_analysis/dataset_rooftile_640/**
+/code/data_analysis/dataset_rooftile_small/**
+*.JPG
+code/data_analysis/dataset_rooftile_small/d0002_b_nrd_g_VH/d0002_b_nrd_g_VH_1699026003_DSC00658.JPG
```
Comment (Phaneron97) [inline .gitignore]: why does file single file need to be ignored? We shouldn't bloat the .gitignore but instead move everything dataset related to a single folder which will be ignored.

Comment (Bastiaan27) [reply .gitignore]: I agree, please ignore the whole folder with pictures instead of individual ones

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+                # Calculate the position to paste the resized image on the canvas
+                x_offset = (desired_width - resized_image.shape[1]) // 2
+                y_offset = (desired_height - resized_image.shape[0]) // 2
+
+                # Paste the resized image onto the canvas
+                canvas[y_offset:y_offset+resized_image.shape[0], x_offset:x_offset+resized_image.shape[1]] = resized_image
+
+                # Save the rescaled image with the new filename (in the new folder)
+                cv2.imwrite(write_path, canvas)
```
Comment (Bastiaan27) [inline code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: You have a lot of if's and two for loops in this method, this is a lot of code for a single method
It would be best to split code up into multiple methods which also makes debugging go smoother

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            if resolution_original[0] != target_size[0] or resolution_original[1] != target_size[1]:
+                # Output the current image being processed
+                print(f"Processing image: {image_path}")
+                original_image = cv2.imread(image_path)
+
+                # if image is verticl, rotate it so all images are horizontal.
+                if resolution_original[0] < resolution_original[1]:
+                    print(f'resolution was {resolution_original}, so rotated it')
+                    original_image = cv2.rotate(original_image, cv2.ROTATE_90_COUNTERCLOCKWISE)
+                    resolution_original = (original_image.shape[1], original_image.shape[0])
```
Comment (Bastiaan27) [inline code/data_analysis/data_analysis_resize_all_keep_ratio.py:64]: Since you resize and keep the ratio is it really necessary to rotate the pictures?
This would maybe add a bias to the AI if we change the rotation (so it will be more accurate when a rooftile is laying horizontal?)

**Review by Bastiaan27 — CHANGES_REQUESTED** [review]
I like what you have done, it still needs some changes but I really like what you've done so far.

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py (hunk 2 above)
Comment (Bastiaan27) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: As a general guideline I would advise to try and keep a method under 30 lines of code

File: .gitignore (hunk 1 above)
Comment (Sara-Eftekhar-Azam) [reply .gitignore]: This was in an older commit. I changed in in the next commit. Please see the latest file

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py (hunk 2 above)
Comment (Sara-Eftekhar-Azam) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: I you don't count the comments and empty lines, it is under 30 lines of code. Breaking code into other methods is not always possible. I will try to look at it and break it into more methods, though.

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py (hunk 3 above)
Comment (Sara-Eftekhar-Azam) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:64]: I see your point. I will try to change it so it does not have to do that.

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py (hunk 2 above)
Comment (Bastiaan27) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: It also enhances the readability, for me right now it is hard to see what everything does without having to thoroughly read through the code. 
When you are going to split it up into multiple methods I would advise on giving them clear names so you can immediately understand what it does (for example: make_images_square(), add_black_bars(), check_if_directory_exists() etc.)

Comment (Bastiaan27) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: You are doing great work, I understand that it's not always possible but please keep it in mind

Comment (Sara-Eftekhar-Azam) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: I see your point. I will try to make more methods to make it more clear. Thanks for the feedback.

**Comment by Bastiaan27** [comment]
Please resolve the conversations only when you updated the code so you can still see the conversation without having to fold it out

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py (hunk 2 above)
Comment (Phaneron97) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: a few suggestions to make new functtions:

* ` # Calculate the position to paste the resized image on the canvas
                x_offset = (desired_width - resized_image.shape[1]) // 2
                y_offset = (desired_height - resized_image.shape[0]) // 2`

this can be a new function like `calc_offset` or something.

* 2 `if`s can be their own function
* `                desired_width = target_size[0]
                desired_height = target_size[1]
                width = resolution_original[0]
                height = resolution_original[1]
                ratio = min(desired_width / width, desired_height / height)` 
this can also be its own f unction, something like `get_values` maybe?

Comment (Sara-Eftekhar-Azam) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:86]: Thanks for the suggestions. I will apply them.

File: code/data_analysis/data_analysis_resize_all_keep_ratio.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+                    
+                desired_width = target_size[0]
+                desired_height = target_size[1]
+                width = resolution_original[0]
+                height = resolution_original[1]
+                ratio = min(desired_width / width, desired_height / height)
```
Comment (Bastiaan27) [inline code/data_analysis/data_analysis_resize_all_keep_ratio.py:70]: Why did you rename these values here?
Is a rename needed? Or can you at least rename them when receiving them in the method? Or at the beginning of the method?

Comment (Sara-Eftekhar-Azam) [reply code/data_analysis/data_analysis_resize_all_keep_ratio.py:70]: I wanted to make them shorter. I have changed this part in my new code.

**Review by Sara-Eftekhar-Azam — COMMENTED** [review]
I have made more methods to make it more readable. I also added an option to make horizantal or not. Also, I added an option to use random background colors.

**Comment by Phaneron97** [comment]
[SaraExplanation.docx](https://github.com/Phaneron97/BDA_proj_rooftiles/files/14551391/SaraExplanation.docx)

Gives a clear explanation on the code. squashing and merging this.
